Computed regions appear incorrect in fullscreen

mwyrzykowski opened the pull request:
#### c72a5407c8e91c7a3dec6815a922816423ccbfd4
<pre>
Computed regions appear incorrect in fullscreen
<a href="https://bugs.webkit.org/show_bug.cgi?id=244056">https://bugs.webkit.org/show_bug.cgi?id=244056</a>
&lt;rdar://97562328&gt;

Reviewed by Tim Nguyen.

When in fullscreen, only allow elements under the fullscreen
element to be non-inert. That effectively disables them
from interaction during fullscreen.

* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):

Canonical link: <a href="https://commits.webkit.org/253803@main">https://commits.webkit.org/253803@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a39b1586a1ce93ddb6b19794421b6126c1364a31

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87077 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31163 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17925 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96047 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149657 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91053 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29528 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25796 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91104 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92693 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23844 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73897 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/79210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78828 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79099 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27257 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/12937 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27200 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13951 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2683 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28884 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36800 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28825 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33228 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -60,6 +62,20 @@ static CursorType cursorTypeForElement(Element& element)
     return cursorType;
 }
 
+static bool skipNonFullscreenElements(const RenderObject& regionRenderer)
```
This function name reads like an action ("please skip") but really is a query ("should skip?")

hortont424 commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -60,6 +62,20 @@ static CursorType cursorTypeForElement(Element& element)
+    if (!fullscreenNode)
+        return false;
+
+    for (auto* node = regionRenderer.node(); node; node = node->parentNode()) {
```
Probably could use one of the ElementAncestorIterator functions here instead (doubly so since `currentFullscreenElement` will only give you Elements). But also I expect we already have a "is an ancestor of" function somewhere.

Beyond that, though, it seems unfortunate to walk back up the tree for every region.

Also noticed the `ContainsFullScreenElement` Node bit, but that's not quite what you want (you're looking for "is child of full screen element", right?), sadly.

You should ask around! Someone might have a good idea.

mwyrzykowski replied:
oh maybe I could use `ContainsFullScreenElement` if I stop the building of interaction regions at a higher level rather than from this function

mwyrzykowski commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -60,6 +64,22 @@ static CursorType cursorTypeForElement(Element& element)
+    if (!fullscreenElement)
+        return false;
+
+    for (auto& parentElement : lineageOfType<Element>(element)) {
```
going to see if there is a better way to do this

nt1m reviewed (commented):
I think we should try to leverage inert for the fullscreen API here, the same way we leverage inert for modal dialogs.

You'd simply do this:
https://searchfox.org/wubkat/rev/340573c924b370891b7a0794ef39d4995d3670ee/Source/WebCore/style/StyleAdjuster.cpp#580-581,590-592

but with `document().fullscreenManager().currentFullscreenElement()` instead of `document().activeModalDialog()`.

and invalidate style in the right places if needed.

nt1m commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -67,6 +67,10 @@
 #include <wtf/cocoa/RuntimeApplicationChecksCocoa.h>
 #endif
 
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
```
Do we need to gate this and the other stuff behind the flag?

mwyrzykowski replied:
I'm concerned about potential regressions when the feature is disabled so that is why I gated it

nt1m replied:
I think this particular change is non controversial without the flag. If there are regressions, we should fix them, because we will eventually do this anyway for the new fullscreen API

mwyrzykowski replied:
Ok sounds good, I'll remove the gating

nt1m commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -581,6 +585,10 @@ void Adjuster::adjust(RenderStyle& style, const RenderStyle* userAgentAppearance
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+        if (m_document.fullscreenManager().fullscreenElement())
+            return element == m_document.documentElement();
+#endif
```
Can we be consistent with the `activeModalDialog()` check above (and potentially combine them)?

nt1m commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -591,6 +599,11 @@ void Adjuster::adjust(RenderStyle& style, const RenderStyle* userAgentAppearance
             style.setEffectiveInert(false);
 
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+        if (m_element == m_document.fullscreenManager().fullscreenElement())
```
We do want to make fullscreen elements with the inert attribute inert as well. So ` && !hasInertAttribute(m_element)` is needed.

mwyrzykowski committed:
Computed regions appear incorrect in fullscreen
https://bugs.webkit.org/show_bug.cgi?id=244056
<rdar://97562328>

Reviewed by Tim Nguyen.

When in fullscreen, only allow elements under the fullscreen
element to be non-inert. That effectively disables them
from interaction during fullscreen.

* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):

Canonical link: https://commits.webkit.org/253803@main
